[BAU-297] docs: add contribution guide page (#1476)
contentful/forma-36: packages/forma-36-website/src/content/guidelines/contributing.mdx

---
title: 'Contributing to Forma 36'
slug: '/guidelines/contributing/'
---

Forma 36 is an open-source design system by Contentful and we welcome contributions from our community.

> If we have an idea that could help other users in this page we will teach you how to do it.

Reviewer: ```suggestion
If you have an idea that could help other users in this page we will teach you how to do it.
```

> If you have an idea that could help other users in this page we will teach you how to do it.

Reviewer 2: ```suggestion
If you have an idea that could help other users, a new solution, fix to existing components, or a new component, on this page we will teach you how to transform those ideas into contributions.
```

## Table of contents

- [Before contributing](#before-contributing)
  - [Check if your idea is already present](#check-if-your-idea-is-already-present)
  - [Could an existing component be extended to realize your idea?](#could-an-existing-component-be-extended-to-realize-your-idea)
- [Contribution Process](#contribution-process)
  - [Step 1: Open a Github issue](#step-1-open-a-github-issue)
  - [Step 2: Proposal Review](#step-2-proposal-review)
  - [Step 3: Open a Pull Request](#step-3-open-a-pull-request)
  - [Step 4: Code & Design Review](#step-4-code--design-review)
  - [Step 5: Merge the changes to the main branch](#step-5-merge-the-changes-to-the-main-branch)
- [Bugs & Light contributions](#bugs--light-contributions)
  - [Examples of light contributions](#examples-of-light-contributions)

## Before contributing

### Check if your idea is already present

Before writing a proposal or any code, we ask you to go around this website or our Github repo and

> check if your idea is not yet implemented.

Reviewer 3: Maybe this can be more specific? "Search this website" or something. It can be easy to lose track of what already exists, especially on GitHub.

Author: but I'm saying "look around the website and Github"

I think I didn't understand the suggestion, could you re-write this part so I can see what you have in mind? 🙏

Reviewer 3: If you're completely new you probably won't know what to look for or where, so I think "we ask you to go around this website or our Github repo" can be more actionable. For instance it could be something like "Use the search bar to the left" (or something better 😅 )

Reviewer 4: I would prefer to keep both suggestions, the website and github repo (with a link if possible), because there will be users that would prefer checking the documentation and users that would prefer to go into the codebase and check it by themselves.

Reviewer: maybe add a direct link in the text to the github repo.
has anyone access to our community slack? we could also ask them to go into the community slack forma-36 channel and ask questions there if they are unsure if their idea is yet implemented?

> It’s also a good idea to take a look on the current open proposals,

Reviewer 2: ```suggestion
It’s also a good idea to take a look at the current open [proposals in GitHub](https://github.com/contentful/forma-36/issues?q=is%3Aopen+is%3Aissue+label%3Aproposal),
```

maybe there is something similar to your idea there and you could work together with that proposer.

### Could an existing component be extended to realize your idea?

During your research, did you find a component that almost does what you have in mind?

> If yes, consider proposing an update or refactor of this component.

Reviewer: ```suggestion
If yes, consider proposing an update to e.g. add a new variation or a refactoring of this component.
```

Author: "an update to e.g."
I didn't understand this, I'm sorry
could explain what's the meaning?

Reviewer: i somehow wanted to make clear what all could be considered an update. Maybe this could be clearer:
```suggestion
If yes, consider proposing an update or refactor of this component. This could be a new variation or adding a new property to the component.
```

## Contribution Process

> Now that you checked the library and know if you want to create a new component

Reviewer 2: ```suggestion
Now that you checked the library and you know that you want to create a new component
```

or extend an existing one it’s time to go through the contribution process.

### Step 1: Open a Github issue

The first step is to [open a new issue in Github](https://github.com/contentful/forma-36/issues/new/choose) and choose the correct template.
Remember to describe your idea in detail and maybe use images to support it.

### Step 2: Proposal Review

Forma 36’s internal team checks the issues in the repo regularly and we will provide you with feedback to all proposals.
After some discussion we will approve the proposal and you can work on it. We will help you in anything you need.

It’s a good idea to look into the other components’ code
and our [Code Style Guide](https://github.com/contentful/forma-36/blob/forma-v4/docs/code-style-guide.md) to write your code following our convetions.

> ### Step 3: Open a Pull Request

Reviewer 2: Maybe it's worth adding info around `feat` vs `fix` and how it would be reflected in the lib. so `minor` vs `fix`? 
Dunno, maybe also add that major, breaking changes are closed for external contributions? (I guess they are?)

Author: I prefer to keep it short
The versioning method we seems to be a detail that the contributors don't need to know
or if they want to know they can ask us

I just don't want this to be very long and because it's long people will avoid it

> Once the code is ready, open a Pull Request in our repository.

Reviewer: maybe we should also write that there is a template when opening the PR that will help them figuring out if their proposal is feature-complete?

Author: Yep! I will mention that
thanks 🙌

### Step 4: Code & Design Review

We will review both the code and the design changes (if you have any).
During this step we just want to make sure the solution is maintainable and scalable enough.

### Step 5: Merge the changes to the main branch

After the Pull Request is approved, you can merge the code to the main branch
and it will automatically publish a new version of Forma 36 with your changes.

## Bugs & Light contributions

Sometimes it’s possible to skip steps 1 and 2 of the contribution process.
This is the case for bugs, if you find something that is broken you can open directly a Pull Request with the fix.
This is also true for what we call "light" contributions. They are visual defects, small visual changes, or adding documentation.

### Examples of light contributions

- [Bug - Visited links inside Note component are yellow](https://github.com/contentful/forma-36/issues/866)
- [Proposal - Remove underline in links](https://github.com/contentful/forma-36/issues/627)

> - [Proposal - Change the card border radius to 3 or 4](https://github.com/contentful/forma-36/issues/556)

Reviewer 2: I really liked that visualization that you showed some time ago regarding contribution (like: components exists YES/NO, if YES, do that, if NO do that), maybe we could have it here as well? Or it's an overkill?

Author: Yep! I want to draw a new one

more simplified and then I will put here
Similar to what they are doing here https://hds.hel.fi/contributing/before-contributing#contribution-process

Reviewer 2: Exactly, that would be amazing! 👏

If yes, consider proposing an update or refactor of this component. This could be a new variation or adding a new property to the component.

## Contribution Process

Now that you checked the library and you know that you want to create a new component
or extend an existing one it’s time to go through the contribution process.

### Step 1: Open a Github issue

The first step is to [open a new issue in Github](https://github.com/contentful/forma-36/issues/new/choose) and choose the correct template.
Remember to describe your idea in detail and maybe use images to support it.

### Step 2: Proposal Review

Forma 36’s internal team checks the issues in the repo regularly and we will provide you with feedback to all proposals.
After some discussion we will approve the proposal and you can work on it. We will help you in anything you need.

It’s a good idea to look into the other components’ code
and our [Code Style Guide](https://github.com/contentful/forma-36/blob/forma-v4/docs/code-style-guide.md) to write your code following our convetions.

### Step 3: Open a Pull Request

Once the code is ready, open a Pull Request in our repository.

> The PR template has a template that will provide more information on what to write in the request.

Author: Thanks for the review! I will make the changes

### Step 4: Code & Design Review

We will review both the code and the design changes (if you have any).
During this step we just want to make sure the solution is maintainable and scalable enough.

### Step 5: Merge the changes to the main branch

After the Pull Request is approved, you can merge the code to the main branch
and it will automatically publish a new version of Forma 36 with your changes.

## Bugs & Light contributions

Sometimes it’s possible to skip steps 1 and 2 of the contribution process.

> This is the case for bugs, if you find something that is broken you can open directly a Pull Request with the fix.

Reviewer 5: ```suggestion
For example, for bugs, you can directly open a Pull Request with the fix.
```
